extensions/charts/rke2template.go

```diff
@@ -0,0 +1,139 @@
+package charts
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/shepherd/clients/rancher"
+	"github.com/rancher/shepherd/extensions/cloudcredentials"
+	"github.com/rancher/shepherd/extensions/clusters"
+	"github.com/rancher/shepherd/extensions/defaults"
+	"github.com/rancher/shepherd/extensions/projects"
+	"github.com/rancher/shepherd/pkg/api/steve/catalog/types"
+	"github.com/rancher/shepherd/pkg/namegenerator"
+	"github.com/rancher/shepherd/pkg/wait"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	fleetNamespace = "fleet-default"
+	rke2Provider   = "rke2"
+	localCluster   = "local"
+)
+
+// InstallRKE2TemplateChart installs a rke2 template from a repo.
+func InstallRKE2TemplateChart(client *rancher.Client, repoName, templateName, k8sVersion string, credentials *cloudcredentials.CloudCredential) error {
+	latestVersion, err := client.Catalog.GetLatestChartVersion(templateName, repoName)
+	if err != nil {
+		return err
+	}
+
+	project, err := projects.GetProjectByName(client, localCluster, "System")
+	if err != nil {
+		return err
+	}
+
+	installOptions := &InstallOptions{
+		Cluster: &clusters.ClusterMeta{
+			ID: localCluster,
+		},
+		Version:   latestVersion,
+		ProjectID: project.ID,
+	}
+
+	serverSetting, err := client.Management.Setting.ByID(serverURLSettingID)
+	if err != nil {
+		return err
+	}
+
+	registrySetting, err := client.Management.Setting.ByID(defaultRegistrySettingID)
+	if err != nil {
+		return err
+	}
+
+	chartInstallActionPayload := &payloadOpts{
+		InstallOptions:  *installOptions,
+		Name:            templateName,
+		Namespace:       fleetNamespace,
+		Host:            serverSetting.Value,
+		DefaultRegistry: registrySetting.Value,
+	}
+
+	chartValues, err := client.Catalog.GetChartValues(repoName, templateName, installOptions.Version)
+	if err != nil {
+		return err
+	}
+
+	chartInstallAction := RKE2TemplateInstallAction(chartInstallActionPayload, repoName, templateName, credentials.ID, k8sVersion, fleetNamespace, chartValues)
+
+	catalogClient, err := client.GetClusterCatalogClient(installOptions.Cluster.ID)
+	if err != nil {
+		return err
+	}
+
+	err = client.Catalog.InstallChart(chartInstallAction, repoName)
+	if err != nil {
+		return err
+	}
+
+	client.Session.RegisterCleanupFunc(func() error {
+		err := client.Catalog.UninstallChart(templateName, fleetNamespace, newChartUninstallAction())
+		if err != nil {
+			return err
+		}
+
+		watchAppInterface, err := catalogClient.Apps(fleetNamespace).Watch(context.TODO(), metav1.ListOptions{
+			FieldSelector:  "metadata.name=" + templateName,
+			TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = wait.WatchWait(watchAppInterface, func(event watch.Event) (ready bool, err error) {
```

Comment:
couldn't this use `wait.ResourceDelete(ion)`?

Reply:
Good point.